nginx/ngx_http_js_module.c

```diff
@@ -4426,22 +4646,2897 @@ ngx_js_http_init(njs_vm_t *vm)
+
+            if (rc == NGX_ERROR) {
+                return NGX_ERROR;
+            }
+
+            if (!pending && rc == NGX_AGAIN) {
+                ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+                              "async operation inside \"%V\" body filter",
+                              &jlcf->body_filter);
+                return NGX_ERROR;
+            }
+
+            ctx->buf->pos = ctx->buf->last;
+
+        } else {
+            cl = ngx_alloc_chain_link(c->pool);
+            if (cl == NULL) {
+                return NGX_ERROR;
+            }
+
+            cl->buf = b;
+
+            *ctx->last_out = cl;
+            ctx->last_out = &cl->next;
+        }
+
+        in = in->next;
+    }
+
+    return NGX_OK;
+}
+
+
+static ngx_http_request_t *
+ngx_http_qjs_request(JSValueConst val)
+{
+    ngx_http_qjs_request_t  *req;
+
+    req = JS_GetOpaque(val, NGX_QJS_CLASS_ID_HTTP_REQUEST);
+    if (req == NULL) {
+        return NULL;
+    }
+
+    return req->request;
+}
+
+
+static JSValue
+ngx_http_qjs_request_make(JSContext *cx, ngx_int_t proto_id,
+    ngx_http_request_t *r)
+{
+    int                      ret;
+    JSValue                  request;
+    ngx_http_qjs_request_t  *req;
+
+    request = JS_NewObjectClass(cx, proto_id);
+    if (JS_IsException(request)) {
+        return JS_EXCEPTION;
+    }
+
+    ret = qjs_set_to_string_tag(cx, request, "Request");
+    if (ret == -1) {
+        JS_FreeValue(cx, request);
+        return JS_EXCEPTION;
+    }
+
+    req = js_malloc(cx, sizeof(ngx_http_qjs_request_t));
+    if (req == NULL) {
+        return JS_ThrowOutOfMemory(cx);
+    }
+
+    req->request = r;
+    req->args = JS_UNDEFINED;
+    req->request_body = JS_UNDEFINED;
+    req->response_body = JS_UNDEFINED;
+
+    JS_SetOpaque(request, req);
+
+    return request;
+}
+
+
+static void
+ngx_http_qjs_request_finalizer(JSRuntime *rt, JSValue val)
+{
+    ngx_http_qjs_request_t  *req;
+
+    req = JS_GetOpaque(val, NGX_QJS_CLASS_ID_HTTP_REQUEST);
+    if (req == NULL) {
+        return;
+    }
+
+    JS_FreeValueRT(rt, req->args);
+    JS_FreeValueRT(rt, req->request_body);
+    JS_FreeValueRT(rt, req->response_body);
+
+    js_free_rt(rt, req);
+}
+
+
+static ngx_engine_t *
+ngx_engine_qjs_clone(ngx_js_ctx_t *ctx, ngx_js_loc_conf_t *cf,
+    njs_int_t proto_id, void *external)
+{
+    JSValue             proto;
+    JSContext          *cx;
+    ngx_engine_t       *engine;
+    ngx_http_js_ctx_t  *hctx;
+
+    engine = ngx_qjs_clone(ctx, cf, external);
+    if (engine == NULL) {
+        return NULL;
+    }
+
+    cx = engine->u.qjs.ctx;
+
+    if (!JS_IsRegisteredClass(JS_GetRuntime(cx),
+                              NGX_QJS_CLASS_ID_HTTP_REQUEST))
+    {
+        if (JS_NewClass(JS_GetRuntime(cx), NGX_QJS_CLASS_ID_HTTP_REQUEST,
+                        &ngx_http_qjs_request_class) < 0)
+        {
+            return NULL;
+        }
+
+        proto = JS_NewObject(cx);
```

Comment:
Looks like to fail early, we need add after each JS_NewObject() here:

      if (JS_IsException(proto)) {
          return NULL;
      }